// @airtable/blocks/dist/cjs/ui/system/spacing/margin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.marginPropTypes = exports.margin = exports.config = void 0;

var _core = require("@styled-system/core");

var _ensure_numbers_are_within_scale = _interopRequireDefault(require("../utils/ensure_numbers_are_within_scale"));

var _create_style_prop_types = _interopRequireDefault(require("../utils/create_style_prop_types"));

/** @module @airtable/blocks/ui/system: Spacing */

/** */
var config = {
  margin: {
    property: 'margin',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'margin',
      shouldAllowNegativeNumbers: true
    })
  },
  marginTop: {
    property: 'marginTop',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginTop',
      shouldAllowNegativeNumbers: true
    })
  },
  marginRight: {
    property: 'marginRight',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginRight',
      shouldAllowNegativeNumbers: true
    })
  },
  marginBottom: {
    property: 'marginBottom',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginBottom',
      shouldAllowNegativeNumbers: true
    })
  },
  marginLeft: {
    property: 'marginLeft',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginLeft',
      shouldAllowNegativeNumbers: true
    })
  },
  marginX: {
    properties: ['marginLeft', 'marginRight'],
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginX',
      shouldAllowNegativeNumbers: true
    })
  },
  marginY: {
    properties: ['marginTop', 'marginBottom'],
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'marginY',
      shouldAllowNegativeNumbers: true
    })
  }
};
exports.config = config;
var margin = (0, _core.system)(config);
exports.margin = margin;
var marginPropTypes = (0, _create_style_prop_types.default)(margin.propNames);
exports.marginPropTypes = marginPropTypes;
