// @airtable/blocks/dist/cjs/ui/system/spacing/padding.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.paddingPropTypes = exports.padding = exports.config = void 0;

var _core = require("@styled-system/core");

var _ensure_numbers_are_within_scale = _interopRequireDefault(require("../utils/ensure_numbers_are_within_scale"));

var _create_style_prop_types = _interopRequireDefault(require("../utils/create_style_prop_types"));

/** @module @airtable/blocks/ui/system: Spacing */

/** */
var config = {
  padding: {
    property: 'padding',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'padding'
    })
  },
  paddingTop: {
    property: 'paddingTop',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingTop'
    })
  },
  paddingRight: {
    property: 'paddingRight',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingRight'
    })
  },
  paddingBottom: {
    property: 'paddingBottom',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingBottom'
    })
  },
  paddingLeft: {
    property: 'paddingLeft',
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingLeft'
    })
  },
  paddingX: {
    properties: ['paddingLeft', 'paddingRight'],
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingX'
    })
  },
  paddingY: {
    properties: ['paddingTop', 'paddingBottom'],
    scale: 'space',
    transform: (0, _ensure_numbers_are_within_scale.default)({
      propertyName: 'paddingY'
    })
  }
};
exports.config = config;
var padding = (0, _core.system)(config);
exports.padding = padding;
var paddingPropTypes = (0, _create_style_prop_types.default)(padding.propNames);
exports.paddingPropTypes = paddingPropTypes;
