// @airtable/blocks/dist/cjs/event_tracker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trackEvent = trackEvent;
exports.RecordActionStatus = void 0;

var _get_sdk = _interopRequireDefault(require("./get_sdk"));

/** @hidden */
function trackEvent(eventSchemaName) {
  var eventData = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  (0, _get_sdk.default)().__airtableInterface.trackEvent(eventSchemaName, eventData);
}
/**
 * Used for logging blockInstallation.performRecordAction event
 * Mirrored from hyperbase block_installation_record_action_status.tsx
 *
 * @hidden
 * */


var RecordActionStatus;
exports.RecordActionStatus = RecordActionStatus;

(function (RecordActionStatus) {
  RecordActionStatus["SUCCESS"] = "SUCCESS";
  RecordActionStatus["IGNORED"] = "IGNORED";
  RecordActionStatus["SETTINGS_INVALID"] = "SETTINGS_INVALID";
  RecordActionStatus["TABLE_MISMATCH"] = "TABLE_MISMATCH";
})(RecordActionStatus || (exports.RecordActionStatus = RecordActionStatus = {}));
