// @airtable/blocks/dist/cjs/ui/geometry/rect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _point = _interopRequireDefault(require("./point"));

/** @internal */
var Rect =
/*#__PURE__*/
function () {
  /** @internal */

  /** @internal */

  /** @internal */

  /** @internal */

  /** @internal */
  function Rect(x, y, width, height) {
    (0, _classCallCheck2.default)(this, Rect);
    (0, _defineProperty2.default)(this, "x", void 0);
    (0, _defineProperty2.default)(this, "y", void 0);
    (0, _defineProperty2.default)(this, "width", void 0);
    (0, _defineProperty2.default)(this, "height", void 0);
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
  }
  /** @internal */


  (0, _createClass2.default)(Rect, [{
    key: "growBy",
    value: function growBy(amount) {
      return new Rect(this.x - amount, this.y - amount, this.width + 2 * amount, this.height + 2 * amount);
    }
    /** @internal */

  }, {
    key: "top",
    value: function top() {
      return this.y;
    }
    /** @internal */

  }, {
    key: "left",
    value: function left() {
      return this.x;
    }
    /** @internal */

  }, {
    key: "right",
    value: function right() {
      return this.x + this.width;
    }
    /** @internal */

  }, {
    key: "bottom",
    value: function bottom() {
      return this.y + this.height;
    }
    /** @internal */

  }, {
    key: "centerPoint",
    value: function centerPoint() {
      return new _point.default(this.x + this.width / 2, this.y + this.height / 2);
    }
  }]);
  return Rect;
}();

exports.default = Rect;
