// @airtable/blocks/dist/cjs/ui/theme/default_theme/text_button_variants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _emotion = require("emotion");

var _tokens = require("./tokens");

var NOT_DISABLED = '&:not([aria-disabled="true"])';
var DISABLED = '&[aria-disabled="true"]';
var baseStyles = (0, _emotion.css)({
  alignItems: 'center',
  justifyContent: 'center',
  borderRadius: _tokens.radii.default,
  outline: 'none',
  fontWeight: _tokens.fontWeights.strong,
  [NOT_DISABLED]: {
    cursor: 'pointer',
    '&:hover': {
      opacity: _tokens.opacities.quiet
    },
    '&:active': {
      opacity: 1
    },
    '&:focus': {
      boxShadow: "0 0 0 2px ".concat(_tokens.colors.darken3)
    }
  },
  [DISABLED]: {
    opacity: _tokens.opacities.quieter
  }
});
var textButtonVariants = {
  default: (0, _emotion.cx)(baseStyles, (0, _emotion.css)({
    color: _tokens.colors.blueBright
  })),
  dark: (0, _emotion.cx)(baseStyles, (0, _emotion.css)({
    color: _tokens.colors.dark
  })),
  light: (0, _emotion.cx)(baseStyles, (0, _emotion.css)({
    color: _tokens.colors.light
  }))
};
var _default = textButtonVariants;
exports.default = _default;
