// @airtable/blocks/dist/cjs/ui/box.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emotion = require("emotion");

var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));

var _index = require("./system/index");

var _aria_props = require("./types/aria_props");

var _tooltip_anchor_props = require("./types/tooltip_anchor_props");

var _data_attributes_prop = require("./types/data_attributes_prop");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * A box component for creating layouts.
 *
 * [[ Story id="box--example" title="Box example" ]]
 *
 * @component
 * @docsPath UI/components/Box
 */
var Box = (props, ref) => {
  var _props$as = props.as,
      Component = _props$as === void 0 ? 'div' : _props$as,
      id = props.id,
      children = props.children,
      className = props.className,
      style = props.style,
      tabIndex = props.tabIndex,
      role = props.role,
      onClick = props.onClick,
      onMouseEnter = props.onMouseEnter,
      onMouseLeave = props.onMouseLeave,
      hasOnClick = props.hasOnClick,
      _props$dataAttributes = props.dataAttributes,
      dataAttributes = _props$dataAttributes === void 0 ? {} : _props$dataAttributes,
      ariaLabel = props['aria-label'],
      ariaLabelledBy = props['aria-labelledby'],
      ariaDescribedBy = props['aria-describedby'],
      ariaControls = props['aria-controls'],
      ariaExpanded = props['aria-expanded'],
      ariaHasPopup = props['aria-haspopup'],
      ariaHidden = props['aria-hidden'],
      ariaLive = props['aria-live'],
      styleProps = (0, _objectWithoutProperties2.default)(props, ["as", "id", "children", "className", "style", "tabIndex", "role", "onClick", "onMouseEnter", "onMouseLeave", "hasOnClick", "dataAttributes", "aria-label", "aria-labelledby", "aria-describedby", "aria-controls", "aria-expanded", "aria-haspopup", "aria-hidden", "aria-live"]);
  var classNameForStyleProps = (0, _use_styled_system.default)(styleProps);
  return React.createElement(Component, (0, _extends2.default)({
    ref: ref,
    id: id,
    className: (0, _emotion.cx)(classNameForStyleProps, className),
    style: style,
    tabIndex: tabIndex,
    role: role // TODO (stephen): remove tooltip anchor props
    ,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-controls": ariaControls,
    "aria-expanded": ariaExpanded,
    "aria-haspopup": ariaHasPopup,
    "aria-hidden": ariaHidden,
    "aria-live": ariaLive
  }, dataAttributes), children);
};

var ForwardedRefBox = React.forwardRef(Box);
ForwardedRefBox.propTypes = _objectSpread({
  as: _propTypes.default.oneOf(['div', 'span', 'section', 'main', 'nav', 'header', 'footer', 'aside', 'article', 'address', 'hgroup', 'blockquote', 'figure', 'figcaption', 'ol', 'ul', 'li', 'pre']),
  id: _propTypes.default.string,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.number,
  role: _propTypes.default.string,
  dataAttributes: _data_attributes_prop.dataAttributesPropType
}, _aria_props.ariaPropTypes, {}, _tooltip_anchor_props.tooltipAnchorPropTypes, {}, _index.allStylesPropTypes);
ForwardedRefBox.displayName = 'Box';
var _default = ForwardedRefBox;
exports.default = _default;
