// @airtable/blocks/dist/cjs/ui/collaborator_token.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.collaboratorTokenStylePropTypes = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emotion = require("emotion");

var React = _interopRequireWildcard(require("react"));

var _core = require("@styled-system/core");

var _box = _interopRequireDefault(require("./box"));

var _text = _interopRequireDefault(require("./text"));

var _baymax_utils = require("./baymax_utils");

var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));

var _system = require("./system");

var _tooltip_anchor_props = require("./types/tooltip_anchor_props");

var _use_base = _interopRequireDefault(require("./use_base"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var UNKNOWN_PROFILE_PIC_URL = 'https://static.airtable.com/images/userIcons/user_icon_unknown.png';
/**
 * Style props for the {@link CollaboratorToken} component. Accepts:
 * * {@link FlexItemSetProps}
 * * {@link MarginProps}
 * * {@link PositionSetProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.flexItemSet, _system.positionSet, _system.margin);

var collaboratorTokenStylePropTypes = _objectSpread({}, _system.flexItemSetPropTypes, {}, _system.positionSetPropTypes, {}, _system.marginPropTypes);
/**
 * Props for the {@link CollaboratorToken} component. Also accepts:
 * * {@link CollaboratorTokenStyleProps}
 *
 * @noInheritDoc
 * @docsPath UI/components/CollaboratorToken
 */


exports.collaboratorTokenStylePropTypes = collaboratorTokenStylePropTypes;

/**
 * A version of `Collaborator` that doesn't connect to base. Not part of the public API.
 *
 * @hidden
 */
var StaticCollaboratorToken = props => {
  var collaborator = props.collaborator,
      _props$isActive = props.isActive,
      isActive = _props$isActive === void 0 ? true : _props$isActive,
      onMouseEnter = props.onMouseEnter,
      onMouseLeave = props.onMouseLeave,
      onClick = props.onClick,
      hasOnClick = props.hasOnClick,
      className = props.className,
      style = props.style,
      styleProps = (0, _objectWithoutProperties2.default)(props, ["collaborator", "isActive", "onMouseEnter", "onMouseLeave", "onClick", "hasOnClick", "className", "style"]);
  var classNameForStyledProps = (0, _use_styled_system.default)(styleProps, styleParser);
  var userName = collaborator.name || collaborator.email || 'Unknown';
  var profilePicUrl = collaborator.profilePicUrl || UNKNOWN_PROFILE_PIC_URL;
  return React.createElement(_box.default // TODO (stephen): remove tooltip anchor props
  , {
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    className: (0, _emotion.cx)((0, _baymax_utils.baymax)('truncate print-color-exact'), classNameForStyledProps, className),
    style: style,
    alignItems: "center",
    display: "inline-flex",
    opacity: isActive ? 1 : 0.5
  }, profilePicUrl && React.createElement(_box.default, {
    className: (0, _baymax_utils.baymax)('background-cover background-center'),
    style: {
      backgroundImage: "url(\"".concat(profilePicUrl, "\")")
    },
    border: "1px solid #eee",
    width: "22px",
    height: "22px",
    borderRadius: "circle",
    zIndex: 1,
    flex: "none",
    backgroundColor: "grayLight2"
  }), React.createElement(_text.default, {
    className: (0, _baymax_utils.baymax)('truncate'),
    paddingRight: 2,
    paddingLeft: 3,
    textColor: "dark",
    display: "inline-flex",
    borderRadius: "circle",
    alignItems: "center",
    flex: "none",
    backgroundColor: "grayLight2",
    marginLeft: "-12px",
    lineHeight: 1.4
  }, userName));
};
/**
 * A component that shows a single collaborator in a small token, to be displayed inline or in a list of choices.
 *
 * [[ Story id="collaboratortoken--example" title="Collaborator token example" ]]
 *
 * @component
 * @docsPath UI/components/CollaboratorToken
 */


var CollaboratorToken = props => {
  // Re-render when collaborator info updates. This is to ensure isActive is accurate.
  var base = (0, _use_base.default)();
  var activeCollaborators = base.activeCollaborators;
  var isActive = activeCollaborators.some(activeCollaborator => {
    return activeCollaborator.id === props.collaborator.id;
  });
  return React.createElement(StaticCollaboratorToken, (0, _extends2.default)({}, props, {
    isActive: isActive
  }));
};

CollaboratorToken.propTypes = _objectSpread({
  collaborator: _propTypes.default.shape({
    id: _propTypes.default.string,
    email: _propTypes.default.string,
    name: _propTypes.default.string,
    profilePicUrl: _propTypes.default.string,
    status: _propTypes.default.string
  }).isRequired,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, _tooltip_anchor_props.tooltipAnchorPropTypes, {}, collaboratorTokenStylePropTypes);
CollaboratorToken.Static = StaticCollaboratorToken;
var _default = CollaboratorToken;
exports.default = _default;
