// @airtable/blocks/dist/cjs/ui/field_picker_synced.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var _get_sdk = _interopRequireDefault(require("../get_sdk"));

var _global_config_synced_component_helpers = _interopRequireDefault(require("./global_config_synced_component_helpers"));

var _field_picker = _interopRequireWildcard(require("./field_picker"));

var _use_synced = _interopRequireDefault(require("./use_synced"));

var _use_watchable = _interopRequireDefault(require("./use_watchable"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * A wrapper around the {@link FieldPicker} component that syncs with {@link GlobalConfig}.
 *
 * [[ Story id="modelpickers--fieldpickersynced-example" title="Synced field picker example" ]]
 *
 * @docsPath UI/components/FieldPickerSynced
 * @groupPath UI/components/FieldPicker
 * @component
 */
var FieldPickerSynced = (props, ref) => {
  var globalConfigKey = props.globalConfigKey,
      _onChange = props.onChange,
      disabled = props.disabled,
      table = props.table,
      restOfProps = (0, _objectWithoutProperties2.default)(props, ["globalConfigKey", "onChange", "disabled", "table"]);

  var _useSynced = (0, _use_synced.default)(globalConfigKey),
      _useSynced2 = (0, _slicedToArray2.default)(_useSynced, 3),
      fieldId = _useSynced2[0],
      setFieldId = _useSynced2[1],
      canSetFieldId = _useSynced2[2];

  (0, _use_watchable.default)((0, _get_sdk.default)().base, ['tables']);
  (0, _use_watchable.default)(table, ['fields']);

  function _getFieldFromGlobalConfigValue() {
    if (!table || table.isDeleted) {
      return null;
    }

    return typeof fieldId === 'string' && table ? table.getFieldByIdIfExists(fieldId) : null;
  }

  return React.createElement(_field_picker.default, (0, _extends2.default)({}, restOfProps, {
    ref: ref,
    table: table,
    field: _getFieldFromGlobalConfigValue(),
    onChange: field => {
      setFieldId(field ? field.id : null);

      if (_onChange) {
        _onChange(field);
      }
    },
    disabled: disabled || !canSetFieldId
  }));
};

var ForwardedRefFieldPickerSynced = React.forwardRef(FieldPickerSynced);
ForwardedRefFieldPickerSynced.displayName = 'FieldPickerSynced';
ForwardedRefFieldPickerSynced.propTypes = _objectSpread({
  globalConfigKey: _global_config_synced_component_helpers.default.globalConfigKeyPropType
}, _field_picker.sharedFieldPickerPropTypes);
var _default = ForwardedRefFieldPickerSynced;
exports.default = _default;
