// @airtable/blocks/dist/cjs/ui/view_picker_synced.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var _get_sdk = _interopRequireDefault(require("../get_sdk"));

var _view_picker = _interopRequireWildcard(require("./view_picker"));

var _global_config_synced_component_helpers = _interopRequireDefault(require("./global_config_synced_component_helpers"));

var _use_synced = _interopRequireDefault(require("./use_synced"));

var _use_watchable = _interopRequireDefault(require("./use_watchable"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * A wrapper around the {@link ViewPicker} component that syncs with {@link GlobalConfig}.
 *
 * [[ Story id="modelpickers--viewpickersynced-example" title="Synced view picker example" ]]
 *
 * @docsPath UI/components/ViewPickerSynced
 * @groupPath UI/components/ViewPicker
 * @component
 */
var ViewPickerSynced = (props, ref) => {
  var globalConfigKey = props.globalConfigKey,
      table = props.table,
      _onChange = props.onChange,
      disabled = props.disabled,
      restOfProps = (0, _objectWithoutProperties2.default)(props, ["globalConfigKey", "table", "onChange", "disabled"]);

  var _useSynced = (0, _use_synced.default)(globalConfigKey),
      _useSynced2 = (0, _slicedToArray2.default)(_useSynced, 3),
      viewId = _useSynced2[0],
      setViewId = _useSynced2[1],
      canSetViewId = _useSynced2[2];

  (0, _use_watchable.default)((0, _get_sdk.default)().base, ['tables']);
  (0, _use_watchable.default)(table, ['views']);

  function _getViewFromGlobalConfigValue() {
    if (!table || table.isDeleted) {
      return null;
    }

    return typeof viewId === 'string' && table ? table.getViewByIdIfExists(viewId) : null;
  }

  return React.createElement(_view_picker.default, (0, _extends2.default)({}, restOfProps, {
    ref: ref,
    table: table,
    view: _getViewFromGlobalConfigValue(),
    onChange: view => {
      setViewId(view ? view.id : null);

      if (_onChange) {
        _onChange(view);
      }
    },
    disabled: disabled || !canSetViewId
  }));
};

var ForwardedRefViewPickerSynced = React.forwardRef(ViewPickerSynced);
ForwardedRefViewPickerSynced.displayName = 'ViewPickerSynced';
ForwardedRefViewPickerSynced.propTypes = _objectSpread({
  globalConfigKey: _global_config_synced_component_helpers.default.globalConfigKeyPropType
}, _view_picker.sharedViewPickerPropTypes);
var _default = ForwardedRefViewPickerSynced;
exports.default = _default;
