// @airtable/blocks/dist/cjs/models/models.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FieldType", {
  enumerable: true,
  get: function get() {
    return _field.FieldType;
  }
});
Object.defineProperty(exports, "ViewType", {
  enumerable: true,
  get: function get() {
    return _view.ViewType;
  }
});
Object.defineProperty(exports, "Base", {
  enumerable: true,
  get: function get() {
    return _base.default;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function get() {
    return _table.default;
  }
});
Object.defineProperty(exports, "Field", {
  enumerable: true,
  get: function get() {
    return _field2.default;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function get() {
    return _view2.default;
  }
});
Object.defineProperty(exports, "Record", {
  enumerable: true,
  get: function get() {
    return _record.default;
  }
});
Object.defineProperty(exports, "RecordQueryResult", {
  enumerable: true,
  get: function get() {
    return _record_query_result.default;
  }
});
Object.defineProperty(exports, "TableOrViewQueryResult", {
  enumerable: true,
  get: function get() {
    return _table_or_view_query_result.default;
  }
});
Object.defineProperty(exports, "LinkedRecordsQueryResult", {
  enumerable: true,
  get: function get() {
    return _linked_records_query_result.default;
  }
});
exports.recordColoring = void 0;

var recordColoring = _interopRequireWildcard(require("./record_coloring"));

exports.recordColoring = recordColoring;

var _create_aggregators = _interopRequireDefault(require("./create_aggregators"));

var _field = require("../types/field");

var _view = require("../types/view");

var _base = _interopRequireDefault(require("./base"));

var _table = _interopRequireDefault(require("./table"));

var _field2 = _interopRequireDefault(require("./field"));

var _view2 = _interopRequireDefault(require("./view"));

var _record = _interopRequireDefault(require("./record"));

var _record_query_result = _interopRequireDefault(require("./record_query_result"));

var _table_or_view_query_result = _interopRequireDefault(require("./table_or_view_query_result"));

var _linked_records_query_result = _interopRequireDefault(require("./linked_records_query_result"));

/** @ignore */

/** */
var initializedAggregators = null;
Object.defineProperty(exports, 'aggregators', {
  enumerable: true,
  get: () => {
    if (!initializedAggregators) {
      initializedAggregators = (0, _create_aggregators.default)();
    }

    return initializedAggregators;
  }
});
